### PR TITLE
Remove level_source_images image column

### PR DESCRIPTION
**Pull request by laurelfan**
## Move all images to S3

All images theoretically have been moved to s3 but there are a few stragglers:

```
ubuntu@production-console:~/production/dashboard$ ./scripts/archive/move_all_images_to_s3 
moved 0
no level source 547
invalid image 227
blank image 82
other errors 0
already in S3 0
2.718721609s

```
- _no level source_: images are somehow not associated with an actual level source. There is currently no way to reference these images on the site (the ids used in the image urls are the LevelSource id not the LevelSourceImage id).
- _invalid image_: imagemagick could not process these images so we didn't upload them to S3. these are all uploaded from the client so I would blame these on flaky clients.
- _blank image_: the "image" field is empty or null. not sure how this happened (probably a bug we had and fixed), but when we implemented the gallery, we had a bunch of these on common LevelSources that I fixed by replacing the existing blank images with new client uploaded images (normally we don't save a new image when there is an existing one). Likely these are some leftovers on solutions that nobody has created in the last year.

I think these can all be deleted. I could be convinced to upload the invalid images if someone else has a strong opinion in favor of doing that.
## Update the code to ignore the image column

See this PR.
## DB refactor step 1: remove the image column

Current schema:

```
mysql> desc level_source_images;
+-----------------+------------+------+-----+---------+----------------+
| Field           | Type       | Null | Key | Default | Extra          |
+-----------------+------------+------+-----+---------+----------------+
| id              | int(11)    | NO   | PRI | NULL    | auto_increment |
| level_source_id | int(11)    | YES  | MUL | NULL    |                |
| image           | mediumblob | YES  |     | NULL    |                |
| created_at      | datetime   | YES  |     | NULL    |                |
| updated_at      | datetime   | YES  |     | NULL    |                |
+-----------------+------------+------+-----+---------+----------------+
5 rows in set (0.00 sec)

mysql> desc level_sources;
+------------+----------------+------+-----+---------+----------------+
| Field      | Type           | Null | Key | Default | Extra          |
+------------+----------------+------+-----+---------+----------------+
| id         | int(11)        | NO   | PRI | NULL    | auto_increment |
| level_id   | int(11)        | YES  | MUL | NULL    |                |
| md5        | varchar(32)    | NO   |     | NULL    |                |
| data       | varchar(20000) | NO   |     | NULL    |                |
| created_at | datetime       | YES  |     | NULL    |                |
| updated_at | datetime       | YES  |     | NULL    |                |
| hidden     | tinyint(1)     | YES  |     | 0       |                |
+------------+----------------+------+-----+---------+----------------+
7 rows in set (0.00 sec)
```

We can now remove the mediumblob image column from level_source_images.

This should be done in a separate deployment so that the code that uses the column is ensured to be gone before the column is
## ? DB refactor step 2: move level_source_images information to a column in level_sources

The information that this level_source_images is now the single bit "does this level_source have an image" (all the urls and such are generated from information from the LevelSource and the Level, we don't actually do anything with the timestamps in LevelSourceImage). We could move this to a column in the level_source table.

```
mysql> desc level_sources;
+------------+----------------+------+-----+---------+----------------+
| Field      | Type           | Null | Key | Default | Extra          |
+------------+----------------+------+-----+---------+----------------+
| id         | int(11)        | NO   | PRI | NULL    | auto_increment |
| level_id   | int(11)        | YES  | MUL | NULL    |                |
| md5        | varchar(32)    | NO   |     | NULL    |                |
| data       | varchar(20000) | NO   |     | NULL    |                |
| created_at | datetime       | YES  |     | NULL    |                |
| updated_at | datetime       | YES  |     | NULL    |                |
| hidden     | tinyint(1)     | YES  |     | 0       |                |
| image      | tinyint(1)     | YES  |     | 0       |                |
+------------+----------------+------+-----+---------+----------------+
```

### CODE REVIEW

**Comment by laurelfan** [comment]
A bit orthogonal from this... but we may want to fix the permissions on all level source images in s3 for all images that were uploaded while we had the s3 gem permissions problem.

File: dashboard/app/models/level_source_image.rb [hunk 1]
```diff
@@ -8,39 +8,38 @@ class LevelSourceImage < ActiveRecord::Base
...
+        framed_image = ImageLib::overlay_image(:background_url => Rails.root.join(image_filename),
+                                               :foreground_blob => image).to_blob
+      rescue Magick::ImageMagickError # something wrong with the image
+        return false
+      end
+
+      unless AWS::S3.upload_to_bucket('cdo-art', s3_framed_filename, framed_image, no_random: true)
+        return false
+      end
     end
```
Comment (philbogle) [inline dashboard/app/models/level_source_image.rb]: It's not obvious what this returns in the success case, could you add an explicit return true?

Comment (laurelfan) [inline dashboard/app/models/level_source_image.rb]: it returns the return value of self.save... do you mean something like

````
if self.save
  return true
end
````

or
````
return self.save
````

File: dashboard/app/models/level_source_image.rb [hunk 2]
```diff
@@ -8,39 +8,38 @@ class LevelSourceImage < ActiveRecord::Base
...
     unless AWS::S3.upload_to_bucket('cdo-art', s3_filename, image, no_random: true)
       return false
     end
 
-    unless AWS::S3.upload_to_bucket('cdo-art', s3_framed_filename, framed_image, no_random: true)
-      return false
+    # create the framed image
+    if level_source.level.game.app == Game::ARTIST
+      if level_source.level.try(:skin) == 'anna' || level_source.level.try(:skin) == 'elsa'
+        image_filename = "app/assets/images/blank_sharing_drawing_#{level_source.level.skin}.png"
```
Comment (philbogle) [inline dashboard/app/models/level_source_image.rb]: This is the filename for the frame, is that correct?  How about calling this frame_filename to make its role clearer?

File: dashboard/app/models/level_source_image.rb [hunk 3]
```diff
@@ -8,39 +8,38 @@ class LevelSourceImage < ActiveRecord::Base
...
+    # upload original image
     unless AWS::S3.upload_to_bucket('cdo-art', s3_filename, image, no_random: true)
       return false
     end
 
-    unless AWS::S3.upload_to_bucket('cdo-art', s3_framed_filename, framed_image, no_random: true)
-      return false
+    # create the framed image
+    if level_source.level.game.app == Game::ARTIST
+      if level_source.level.try(:skin) == 'anna' || level_source.level.try(:skin) == 'elsa'
```
Comment (philbogle) [inline dashboard/app/models/level_source_image.rb]: Consider moving the code for creating the framed image into a helper function "create_framed_image(image, level)".  Even though this is only called once, I think it makes the code a bit easier to read.

File: dashboard/app/views/gallery_activities/_gallery_activities.html.haml [hunk 4]
```diff
@@ -7,7 +7,7 @@
   %div.gallery_activity
     - activity = gallery_activity.activity
     - level_source = activity.level_source
-    = link_to image_tag(original_image_url(level_source)), level_source
+    = link_to image_tag(level_source.level_source_image.s3_url), level_source
```
Comment (philbogle) [inline dashboard/app/views/gallery_activities/_gallery_activities.html.haml]: Some templates have checks like "if level_source && level_source.level_source_image" but this one doesn't.  Does it need it?

File: dashboard/test/controllers/home_controller_test.rb [hunk 5]
```diff
@@ -68,8 +68,7 @@ def setup_user_with_gallery
     5.times do
       create :gallery_activity,
         user: @user,
-        autosaved: true,
-        activity: create(:activity, user: @user, level: create(:level, game: Game.find_by_app(Game::ARTIST)))
```
Comment (philbogle) [inline dashboard/test/controllers/home_controller_test.rb]: Why does activity go away here?

**Comment by philbogle** [comment]
Done reading. After comments are answered, LGTM.

File: dashboard/app/models/level_source_image.rb (hunk 1 above)
Comment (philbogle) [inline dashboard/app/models/level_source_image.rb]: Sorry, the diff was confusing me and I couldn't read that; the current code is fine.

File: dashboard/app/views/gallery_activities/_gallery_activities.html.haml (hunk 4 above)
Comment (laurelfan) [inline dashboard/app/views/gallery_activities/_gallery_activities.html.haml]: that is checked in line 6 above

File: dashboard/test/controllers/home_controller_test.rb (hunk 5 above)
Comment (laurelfan) [inline dashboard/test/controllers/home_controller_test.rb]: I added it to the factory

**Pull request merged**